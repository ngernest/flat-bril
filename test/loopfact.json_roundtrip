{
  "functions": [
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "input"
          ],
          "dest": "value",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v1"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "value"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.2"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v5",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v4",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "v6"
          ],
          "funcs": [],
          "labels": [
            "for.body.2",
            "for.end.2"
          ],
          "op": "br"
        },
        {
          "label": "for.body.2"
        },
        {
          "args": [
            "result"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v7",
            "v8"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v9"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v11",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v12"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.2"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.2"
        },
        {
          "args": [
            "result"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v13"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v14",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    }
  ]
}
