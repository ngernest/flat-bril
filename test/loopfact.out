@main
	op: id   		dest: value		type: int		args: ["input"]	
	op: const		dest: v1		type: int		value: 1	
	op: id   		dest: result		type: int		args: ["v1"]	
	op: id   		dest: v3		type: int		args: ["value"]	
	op: id   		dest: i 		type: int		args: ["v3"]	
.for.cond.2
	op: id   		dest: v4		type: int		args: ["i"]	
	op: const		dest: v5		type: int		value: 0	
	op: gt   		dest: v6		type: bool		args: ["v4", "v5"]	
	op: br   		args: ["v6"]		labels: ["for.body.2", "for.end.2"]
.for.body.2
	op: id   		dest: v7		type: int		args: ["result"]	
	op: id   		dest: v8		type: int		args: ["i"]	
	op: mul  		dest: v9		type: int		args: ["v7", "v8"]	
	op: id   		dest: result		type: int		args: ["v9"]	
	op: id   		dest: v10		type: int		args: ["i"]	
	op: const		dest: v11		type: int		value: 1	
	op: sub  		dest: v12		type: int		args: ["v10", "v11"]	
	op: id   		dest: i 		type: int		args: ["v12"]	
	op: jmp  		labels: ["for.cond.2"]
.for.end.2
	op: id   		dest: v13		type: int		args: ["result"]	
	op: print		args: ["v13"]	
	op: const		dest: v14		type: int		value: 0	
